src/main/java/de/tum/in/www1/artemis/service/compass/CompassService.java

```diff
@@ -118,38 +121,60 @@ public void removeModelWaitingForAssessment(long exerciseId, long modelSubmissio
     }
 
     /**
+     * Get the (cached) list of models that need to be assessed next. If the number of models in the list is smaller than the configured NUMBER_OF_OPTIMAL_MODELS a new "optimal"
+     * model is added to the list. The models in the list are optimal in the sense of knowledge gain for Compass, helping to automatically assess as many other models as possible.
+     *
      * @param exerciseId the exerciseId
      * @return List of model Ids waiting for an assessment by an assessor
      */
-    public Set<Long> getModelsWaitingForAssessment(long exerciseId) {
+    public List<Long> getModelsWaitingForAssessment(long exerciseId) {
         if (!loadExerciseIfSuspended(exerciseId)) {
-            return new HashSet<>();
+            return new ArrayList<>();
         }
 
-        // TODO: double check that the returned modelSubmissions (respectively their ids) do not have a result yet
+        List<Long> optimalModelIds = compassCalculationEngines.get(exerciseId).getModelsWaitingForAssessment();
+        if (optimalModelIds.size() < NUMBER_OF_OPTIMAL_MODELS) {
```

Comment:
So there is a `number of optimal models`, if we don't have enough models, we add one more.
But why only one? Why does it not try to get more optimal models so that the `number of optimal models is reached`?

Reply:
Good point. The reason we are not loading more optimal models here is that the calculation of the next optimal model done in the `ModelSelector` is rather complex and probably not very fast with a lot of stored submissions. So some users would have to wait until the method was executed 10 times which might negatively affect their user experience. This is just an assumption, though. I wanted to see how much time the corresponding endpoints actually need to respond in production and maybe adjust it later on. An alternative approach is to load the `optimalModelIds` asynchronously, but there were some issues with this in the past.

The current implementation is more like a trade-off between caching some optimal models and decreasing response time. Every time the user navigates to the tutor exercise dashboard one optimal model is loaded. And every time the user actually starts a new assessment one optimal model is loaded and one is removed and sent to the user for the assessment. So in general there should be some optimal models in the "cache".

What do you think about this @krusche ?

Reply:
Definitely do not change it back to async. This was really bad! I think we can definitely get more than one optimal model. I don't think that waiting a few seconds is problematic here. Why don't you fill up the cache with optimal models until NUMBER_OF_OPTIMAL_MODELS is reached?

Reply:
I changed it to fill the cache until `NUMBER_OF_OPTIMAL_MODELS` is reached